feat(pharmacyOrders): TAN-2508: Materialise PharmacyOrderPrescriptions as FhirMedicationRequests

rohan-bes opened the pull request:
### Changes

A pharmacy order is materialised as many FhirMedicationRequests, one for each pharmacyOrderPrescription.
They can be grouped by the `groupIdentifier` field, which links on the pharmacyOrder.id

Most of the complexity in this mapping is in the `dosageInstruction` logic. But it's not too bad :)

### Deploys

- [x] **Deploy to Tamanu Internal** <!-- #deploy -->

### Tests

- [ ] **Run E2E Tests** <!-- #e2e -->

### Remember to...

- ...write or update tests
- ...add UI screenshots and **testing notes** to the Linear issue
- ...add any **manual upgrade steps** to the Linear issue
- ...update the [config reference](https://beyond-essential.slab.com/posts/reference-config-file-0c70ukly), [settings reference](https://beyond-essential.slab.com/posts/reference-settings-0blw1x2q), or any [relevant runbook(s)](https://beyond-essential.slab.com/topics/runbooks-bs04ml6c)
- ...call out additions or changes to **config files** for the deployment team to take note of

<!-- Thank you! -->

Commit:
feat(pharmacyOrders): TAN-2508: Add fhir.medication_requests table

Commit:
feat(pharmacyOrders): TAN-2508: Add getEnumTranslationFunction to TranslatedString model

Commit:
feat(pharmacyOrders): TAN-2508: Add new FHIR types for MedicationRequests

Commit:
feat(pharmacyOrders): TAN-2508: Add FhirMedicationRequest model

Commit:
feat(pharmacyOrders): TAN-2508: Added MedicationRequest to FHIR materialisation config

Commit:
feat(pharmacyOrders): TAN-2508: Added unit test for MedicationRequest materialisation

rohan-bes commented on packages/central-server/__tests__/hl7fhir/materialised/MedicationRequest.test.js:
```diff
@@ -0,0 +1,249 @@
+              },
+              dose: {
+                doseQuantity: {
+                  value: parseInt(prescription.doseAmount),
```
Oh good call!

rohan-bes replied:
Oh wait this is in the tests... less of a good call but sure I'll do it

rohan-bes commented on packages/database/src/utils/fhir/MedicationRequest/getValues.ts:
```diff
@@ -0,0 +1,207 @@
+  const requester = await requesterRef(pharmacyOrder, models);
+  const subject = await subjectRef(pharmacyOrder, models);
+  const encounter = await FhirReference.to(models.FhirEncounter, pharmacyOrder.encounterId);
+  return {
+    lastUpdated: new Date(),
+    identifier: [
+      new FhirIdentifier({
+        system: config.hl7.dataDictionaries.pharmacyOrderPrescriptionId,
+        value: upstream.id,
+      }),
+    ],
+    status: 'active',
+    intent: 'order',
+    groupIdentifier: [
+      new FhirIdentifier({
+        system: config.hl7.dataDictionaries.pharmacyOrderId,
+        value: pharmacyOrder.id,
+      }),
+    ],
+    medication: await medication(upstream, models),
+    authoredOn: pharmacyOrder.createdAt,
+    dosageInstruction: await dosageInstruction(upstream, models),
+    recorder,
+    requester,
+    subject,
+    encounter,
+    note: pharmacyOrder.comments
+      ? [
+          new FhirAnnotation({
+            author: recorder,
+            text: pharmacyOrder.comments,
+          }),
+        ]
+      : null,
+    resolved:
+      requester.isResolved() &&
+      recorder.isResolved() &&
+      subject.isResolved() &&
+      encounter.isResolved(),
+  };
```
Ehh I don't think this is that big of a deal

rohan-bes commented on packages/database/src/utils/fhir/MedicationRequest/getValues.ts:
```diff
@@ -0,0 +1,207 @@
+      new FhirDoseAndRate({
+        dose: {
+          doseQuantity: {
+            value: prescription.doseAmount,
```
Uhh you lie? This comes out as a number in the API

rohan-bes commented on packages/database/src/utils/fhir/MedicationRequest/getValues.ts:
```diff
@@ -0,0 +1,207 @@
+        }),
+      ],
+    }),
+    dispenseRequest: {
+      quantity: prescription.quantity,
+      numberOfRepeatsAllowed: prescription.repeats,
+    },
+  });
```
Oh good call!

Commit:
feat(pharmacyOrders): TAN-2508: Fixed up dispenseRequest column

Commit:
feat(pharmacyOrders): TAN-2508: Updated dbt docs

dannash100 commented on packages/central-server/__tests__/hl7fhir/materialised/MedicationRequest.test.js:
```diff
@@ -0,0 +1,253 @@
+    const [practitioner, patient, drug1, drug2, facility] = await Promise.all([
+      User.create(fake(User)),
+      Patient.create(fake(Patient)),
+      ReferenceData.create({ ...fakeReferenceData(), type: 'drug' }),
+      ReferenceData.create({ ...fakeReferenceData(), type: 'drug' }),
```
Looks like u can do this!
```suggestion
      ReferenceData.create(fakeReferenceData('drug')),
      ReferenceData.create(fakeReferenceData('drug')),
```

Commit:
feat(pharmacyOrders): TAN-2508: Fixed up renamed import after merge

rohan-bes commented on packages/shared/src/services/fhirTypes/timing.js:
```diff
@@ -0,0 +1,47 @@
+      periodUnit: PERIOD_UNITS[random(0, PERIOD_UNITS.length - 1)],
+      timeOfDay: null,
+    });
+  }
```
Nice catch bot!

Commit:
feat(pharmacyOrders): TAN-2508: Fixed fake FhirTiming structure

rohan-bes commented on packages/database/src/utils/fhir/MedicationRequest/getValues.ts:
```diff
@@ -0,0 +1,207 @@
+        dose: {
+          doseQuantity: {
+            value: prescription.doseAmount,
+            unit: prescription.units,
```
Okay so looks like both bugbots are gettting confused by this, I'll switch the dataType of `PrescriptionModel.doseAmount` to NUMBER

Commit:
feat(pharmacyOrders): TAN-2508: Cast doseAmount to number

edmofro reviewed (changes requested):
Cool to see how this stuff works, it's all very declarative which is fun. Don't think I've reviewed a FHIR PR before.

rohan-bes commented on packages/shared/src/services/fhirTypes/quantity.js:
```diff
@@ -0,0 +1,24 @@
+  static SCHEMA() {
+    return yup
+      .object({
+        value: yup.number().nullable().default(null),
```
They are! Often FHIR datatypes there's a few different ways of defining them (for instance here a `quantity` can either be a value/unit definition or a code/system definition). In these cases they often make all the fields optional
https://hl7.org/fhir/R4B/datatypes.html#Quantity

rohan-bes commented on packages/database/src/utils/fhir/MedicationRequest/getQueryToFindUpstreamIds.ts:
```diff
@@ -0,0 +1,155 @@
+import type { Models } from '../../../types/model';
+
+export function getQueryToFindUpstreamIds(models: Models, table: string, id: string) {
```
This is the existing pattern for all the other FHIR resources so I decided to just to follow the pattern here, but overall I'm not much of a fan of this pattern. My main issues with it are:
- Very boilerplatey
- Easy to make mistakes
- Feels like this should be achievable like you say with our existing model level relations

So I'd like to refactor this at some point, not exactly sure how to do that but I'll chuck an item in the Cooldown items

rohan-bes commented on packages/database/src/utils/fhir/MedicationRequest/getValues.ts:
```diff
@@ -0,0 +1,207 @@
+  const orderedByUser = await models.User.findOne({
+    where: { id: pharmacyOrder.orderingClinicianId },
+  });
+  return FhirReference.to(models.FhirPractitioner, pharmacyOrder.orderingClinicianId, {
```
I think I was trying to be a bit clever with these to more confidently ensure that this `id` value would never be undefined (we've seen that in the past in Nauru and it was causing issues). But it doesn't really make much sense, so yep prefer your suggestion :+1:

Commit:
feat(pharmacyOrders): TAN-2508: PR Fixups
- Used cleaner fakeReferenceData call
- Switched reference ids to read more clearly

edmofro commented on packages/shared/src/services/fhirTypes/doseAndRate.js:
```diff
@@ -0,0 +1,51 @@
+        type: FhirCodeableConcept.asYup().nullable().default(null),
+        dose: yup
+          .object({
+            doseQuantity: FhirQuantity.asYup().nullable().default(null),
```
I do agree

edmofro reviewed (approved):
Looks great!

Commit:
feat(pharmacyOrders): TAN-2508: Fixed quotes in dbt docs generation

Commit:
feat(pharmacyOrders): TAN-2508: Add missing fhir triggers

Commit:
Merge branch 'main' into feat/tan-2508-materialise-pharmacy-orders

Commit:
feat(pharmacyOrders): TAN-2508: Updated DBT with new triggers

Commit:
[generate-fake-db]

Commit:
feat(pharmacyOrders): TAN-2508: Reverted generate commit
